pkg/controller/resourceclaim/controller.go

```diff
@@ -0,0 +1,393 @@
+	// objects from the API server. It is shared with other controllers and
+	// therefore the Pod objects in its store should be treated as immutable.
+	podLister v1listers.PodLister
+	podSynced cache.InformerSynced
+
+	// podIndexer has the common PodResourceClaim indexer indexer installed To
+	// limit iteration over pods to those of interest.
+	podIndexer cache.Indexer
+
+	// recorder is used to record events in the API server
+	recorder record.EventRecorder
+
+	queue workqueue.RateLimitingInterface
+}
+
+const (
+	claimKeyPrefix = "claim:"
+	podKeyPrefix   = "pod:"
+)
+
+// NewController creates a ResourceClaim controller.
+func NewController(
+	kubeClient clientset.Interface,
+	podInformer v1informers.PodInformer,
+	claimInformer v1informers.ResourceClaimInformer) (Controller, error) {
+
+	ec := &resourceClaimController{
+		kubeClient:   kubeClient,
+		podLister:    podInformer.Lister(),
+		podIndexer:   podInformer.Informer().GetIndexer(),
+		podSynced:    podInformer.Informer().HasSynced,
+		claimLister:  claimInformer.Lister(),
+		claimsSynced: claimInformer.Informer().HasSynced,
+		queue:        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "resource_claim"),
+	}
+
+	metrics.RegisterMetrics()
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	ec.recorder = eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "resource_claim"})
+
+	if _, err := podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: ec.enqueuePod,
+		UpdateFunc: func(old, updated interface{}) {
+			ec.enqueuePod(updated)
+		},
+		DeleteFunc: ec.enqueuePod,
+	}); err != nil {
+		return nil, err
+	}
+	if _, err := claimInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: ec.onResourceClaimDelete,
+	}); err != nil {
+		return nil, err
+	}
+	if err := ec.podIndexer.AddIndexers(cache.Indexers{podResourceClaimIndex: podResourceClaimIndexFunc}); err != nil {
+		return nil, fmt.Errorf("could not initialize ResourceClaim controller: %w", err)
+	}
+
+	return ec, nil
+}
+
+func (ec *resourceClaimController) enqueuePod(obj interface{}) {
+	if d, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+		obj = d.Obj
+	}
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		// Not a pod?!
+		return
+	}
+
+	if len(pod.Spec.ResourceClaims) == 0 {
+		// Nothing to do for it at all.
+		return
+	}
+
+	// Release reservations of a deleted or completed pod?
+	if pod.DeletionTimestamp != nil && pod.DeletionGracePeriodSeconds != nil && *pod.DeletionGracePeriodSeconds == 0 ||
+		pod.Status.Phase == v1.PodFailed ||
+		pod.Status.Phase == v1.PodSucceeded {
+		for _, podClaim := range pod.Spec.ResourceClaims {
+			claimName := resourceclaim.Name(pod, &podClaim)
+			ec.queue.Add(claimKeyPrefix + pod.Namespace + "/" + claimName)
```

Comment:
It's worth noting that now claims may get processed for which the controller cannot be certain whether the pod was deleted. I am dealing with the ambiguity by maintaining an LRU cache of definitely deleted pods (to avoid hitting the apiserver for all checks) and checking with apiserver if needed.

My current approach with this PR is to put such additional changes into separate "fixup!" commits that I'll squash after review. The older commits still need to be updated directly to adapt to changes in the API, but that's just search/replace.